File: py/numpy/bf/batchnorm.py
import numpy as np

from .module import Module


mean = lambda a: a.mean(1, keepdims=True)


class BatchNormNd(Module):
    def __init__(self, dim, momentum=0.9, eps=1e-5, ndim=None):
        super().__init__()
        self.dim = dim
        self.momentum = momentum
        self.eps = eps
        self.ndim = ndim
        self.gamma = np.ones((dim, 1))
        self.dgamma = np.zeros((dim, 1))
        self.beta = np.zeros((dim, 1))
        self.dbeta = np.zeros((dim, 1))
        self.mov_mean = np.zeros((dim, 1))
        self.mov_std = np.zeros((dim, 1))
        self.x = None

    def each_parameter(self):
        yield self.gamma, self.dgamma
        yield self.beta, self.dbeta

    def forward(self, x, is_t):
        if self.ndim is not None:
            assert x.ndim == self.ndim + 2
        x = x.swapaxes(0, 1)
        s = x.shape
        x = x.reshape(x.shape[0], -1)
        if is_t:
            self.x = x
            x_mean = mean(x)
            x_ctr = x - x_mean
            x_var = mean(x_ctr ** 2)
            x_std = np.sqrt(x_var + self.eps)
            x_norm = x_ctr / x_std
            self.mov_mean = self.momentum * self.mov_mean + \
                (1 - self.momentum) * x_mean
            self.mov_std = self.momentum * self.mov_std + \
                (1 - self.momentum) * x_std
        else:
            x_norm = (x - self.mov_mean) / self.mov_std
        y = x_norm * self.gamma + self.beta
        y = y.reshape(*s)
        return y.swapaxes(0, 1)

    def backward(self, dy):
        dy = dy.swapaxes(0, 1)
        s = dy.shape
        dy = dy.reshape(dy.shape[0], -1)
        x_mean = mean(self.x)
        x_ctr = self.x - x_mean
        x_var = mean(x_ctr ** 2) + self.eps
        x_std = np.sqrt(x_var)
        x_norm = x_ctr / x_std
        self.dbeta = dy.sum(1, keepdims=True)
        self.dgamma = (dy * x_norm).sum(1, keepdims=True)
        dx = dy - mean(dy) - x_ctr / x_var * mean(dy * x_ctr)
        dx = dx * self.gamma / x_std
        dx = dx.reshape(*s)
        return dx.swapaxes(0, 1)


class BatchNorm0d(BatchNormNd):
    def __init__(self, dim, momentum=0.9, eps=1e-5):
        super().__init__(dim, momentum, eps, 0)


class BatchNorm1d(BatchNormNd):
    def __init__(self, dim, momentum=0.9, eps=1e-5):
        super().__init__(dim, momentum, eps, 1)


class BatchNorm2d(BatchNormNd):
    def __init__(self, dim, momentum=0.9, eps=1e-5):
        super().__init__(dim, momentum, eps, 2)


class BatchNorm3d(BatchNormNd):
    def __init__(self, dim, momentum=0.9, eps=1e-5):
        super().__init__(dim, momentum, eps, 3)

File: c/bf/batchnorm.c
#include "bf/batchnorm.h"

#include <assert.h>
#include <math.h>
#include <stdlib.h>

void batchnorm_init(module_t* module, int dim, float momentum, float eps,
                    int ndim) {
    batchnorm_t* f = (batchnorm_t*)module;
    f->type = BATCHNORM;
    f->dim = dim;
    f->ndim = ndim;
    f->momentum = momentum;
    f->eps = eps;
    f->gamma = ones(dim);
    f->dgamma = zeros(dim);
    f->beta = zeros(dim);
    f->dbeta = zeros(dim);
    f->mov_mean = zeros(dim);
    f->mov_std = ones(dim);
    f->x = empty();
}

module_t* batchnorm(int dim, float momentum, float eps, int ndim) {
    module_t* module = (module_t*)malloc(sizeof(batchnorm_t));
    batchnorm_init(module, dim, momentum, eps, ndim);
    return module;
}

void batchnorm_free(module_t* module) {
    batchnorm_t* f = (batchnorm_t*)module;
    tensor_free(f->gamma);
    tensor_free(f->dgamma);
    tensor_free(f->beta);
    tensor_free(f->dbeta);
    tensor_free(f->mov_mean);
    tensor_free(f->mov_std);
    tensor_free(f->x);
    free(f->gamma);
    free(f->dgamma);
    free(f->beta);
    free(f->dbeta);
    free(f->mov_mean);
    free(f->mov_std);
    free(f->x);
}

void batchnorm_zero_grad(module_t* module) {
    batchnorm_t* f = (batchnorm_t*)module;
    tensor_zero(f->dgamma);
    tensor_zero(f->dbeta);
}

tensor_t* batchnorm_forward(module_t* module, tensor_t* x, bool is_t) {
    batchnorm_t* f = (batchnorm_t*)module;
    assert(f->ndim + 2 == x->ndim);
    assert(x->shape[1] == f->dim);
    int batch = x->shape[0];
    int dim = x->shape[1];
    int space = x->size / batch / dim;
    int values_per_dim = x->size / dim;
    tensor_t* y = tensor_clone(x);
    if (is_t) {
        tensor_set(f->x, x);
        for (int d = 0; d < dim; ++d) {
            float x_mean = 0;
            for (int n = 0; n < batch; ++n) {
                for (int s = 0; s < space; ++s) {
                    int idx = (n * dim + d) * space + s;
                    x_mean += x->data[idx];
                }
            }
            x_mean /= values_per_dim;

            float x_var = 0;
            for (int n = 0; n < batch; ++n) {
                for (int s = 0; s < space; ++s) {
                    int idx = (n * dim + d) * space + s;
                    float val = x->data[idx] - x_mean;
                    x_var += val * val;
                }
            }
            x_var /= values_per_dim;
            x_var += f->eps;
            float x_std = sqrt(x_var);

            float gamma = f->gamma->data[d];
            float beta = f->beta->data[d];
            for (int n = 0; n < batch; ++n) {
                for (int s = 0; s < space; ++s) {
                    int idx = (n * dim + d) * space + s;
                    float val = (x->data[idx] - x_mean) / x_std;
                    y->data[idx] = val * gamma + beta;
                }
            }

            f->mov_mean->data[d] = f->momentum * f->mov_mean->data[d] +
                (1 - f->momentum) * x_mean;
            f->mov_std->data[d] = f->momentum * f->mov_std->data[d] +
                (1 - f->momentum) * x_std;
        }
    } else {
        for (int d = 0; d < dim; ++d) {
            float mov_mean = f->mov_mean->data[d];
            float mov_std = f->mov_std->data[d];
            float gamma = f->gamma->data[d];
            float beta = f->beta->data[d];
            for (int n = 0; n < batch; ++n) {
                for (int s = 0; s < space; ++s) {
                    int idx = (n * dim + d) * space + s;
                    float val = (x->data[idx] - mov_mean) / mov_std;
                    y->data[idx] = val * gamma * beta;
                }
            }
        }
    }
    return y;
}

tensor_t* batchnorm_backward(module_t* module, tensor_t* dy) {
    batchnorm_t* f = (batchnorm_t*)module;
    assert(f->ndim + 2 == dy->ndim);
    assert(dy->shape[1] == f->dim);
    int batch = dy->shape[0];
    int dim = dy->shape[1];
    int space = dy->size / batch / dim;
    int values_per_dim = dy->size / dim;
    tensor_t* dx = tensor_clone(dy);
    for (int d = 0; d < dim; ++d) {
        float x_mean = 0;
        for (int n = 0; n < batch; ++n) {
            for (int s = 0; s < space; ++s) {
                int idx = (n * dim + d) * space + s;
                x_mean += f->x->data[idx];
            }
        }
        x_mean /= values_per_dim;

        float x_var = 0;
        for (int n = 0; n < batch; ++n) {
            for (int s = 0; s < space; ++s) {
                int idx = (n * dim + d) * space + s;
                float val = f->x->data[idx] - x_mean;
                x_var += val * val;
            }
        }
        x_var /= values_per_dim;
        x_var += f->eps;
        float x_std = sqrt(x_var);

        float dgamma = 0;
        float dbeta = 0;
        float dy_mean = 0;
        float dy_x_ctr_mean = 0;
        for (int n = 0; n < batch; ++n) {
            for (int s = 0; s < space; ++s) {
                int idx = (n * dim + d) * space + s;
                float dy_val = dy->data[idx];
                float x_ctr = f->x->data[idx] - x_mean;
                float x_norm = x_ctr / x_std;
                dgamma += dy_val * x_norm;
                dbeta += dy_val;
                dy_mean += dy_val;
                dy_x_ctr_mean += dy_val * x_ctr;
            }
        }
        f->dgamma->data[d] += dgamma;
        f->dbeta->data[d] += dbeta;
        dy_mean /= values_per_dim;
        dy_x_ctr_mean /= values_per_dim;

        float gamma = f->gamma->data[d];
        for (int n = 0; n < batch; ++n) {
            for (int s = 0; s < space; ++s) {
                int idx = (n * dim + d) * space + s;
                float dy_val = dy->data[idx];
                float x_ctr = f->x->data[idx] - x_mean;
                float val = dy_val - dy_mean - x_ctr / x_var * dy_x_ctr_mean;
                dx->data[idx] = val * gamma / x_std;
            }
        }
    }
    return dx;
}

void batchnorm_update_step(module_t* module, float lr) {
    batchnorm_t* f = (batchnorm_t*)module;
    tensor_update_step(f->gamma, f->dgamma, lr);
    tensor_update_step(f->beta, f->dbeta, lr);
}
